### PR TITLE
Add `EvidenceVerifier`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
The `EvidenceVerifier` does most of the common verification to be done
on a `Quote3` type and its associated `Collateral`.

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/127" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/127?utm_source=stack-comment).

File: verifier/src/evidence.rs [hunk 1]
```diff
@@ -123,22 +156,175 @@ quote_application_report_body_field_accessor! {
...
 }
 
+/// Verifier for evidence
+///
+/// This will perform most of the verification to be done on [`Evidence`] this includes:
+/// - verifying the certificate chains
+/// - verifying the QE identity
+/// - verifying the TCB info
+/// - verifying the [`TrustedIdentity`] of the application enclave
+/// - verifying the signature of the Quote
```
Comment (nick-mobilecoin) [inline verifier/src/evidence.rs:166]: These aren't here yet, just documenting what I think this type should be doing.

File: verifier/src/evidence.rs [hunk 2]
```diff
@@ -188,168 +374,6 @@ mod test {
...
-                .expect("Failed to parse public key"),
-        )
-        .expect("Failed to decode public key");
-
-        key
-    }
-
-    type VerifierType<T> = And<
-        Quote3Verifier<T>,
-        And<
```
Comment (nick-mobilecoin) [inline verifier/src/evidence.rs:232]: I think the `EvidenceVerifier` IFF it's the way forward removes the main need to try and build up a verifier of all the necessary pieces operating on the `Evidence` type.

I'm still not sure how/where we'll want to handle the Debug flag, https://github.com/mobilecoinfoundation/mobilecoin/issues/3376

File: verifier/src/evidence.rs [hunk 3]
```diff
@@ -432,4 +456,173 @@ mod test {
 
         assert_matches!(Evidence::new(quote, collateral), Err(Error::Serde(_)));
     }
+
+    struct TestDoubleChainVerifier {
+        failed_certificate_subject: String,
+        error: Error,
+    }
```
Comment (nick-mobilecoin) [inline verifier/src/evidence.rs:463]: The `TestDobuleChainVerifier` feels a bit hefty, up for suggestions on minimizing it if possible. I think it's the cost of testing a dependency injection interface.

**Review by eranrund — APPROVED** [review]
Approving since I don't see any deal breakers - only some nits/question regarding SGX stuff I am not familiar with.
Thanks!

File: verifier/src/evidence.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply verifier/src/evidence.rs]: Yes it does, added them now

File: verifier/src/evidence.rs [hunk 4]
```diff
@@ -123,22 +156,175 @@ quote_application_report_body_field_accessor! {
...
+        self.certificate_verifier
+            .verify_certificate_chain(chain, crls, self.time)
+    }
+
+    fn verify_quote_signing_key<Q: AsRef<[u8]>>(
+        &self,
+        quote: &Quote3<Q>,
+        collateral: &Collateral,
+    ) -> Result<VerifyingKey> {
+        let crls = [collateral.root_ca_crl(), collateral.pck_crl()];
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply verifier/src/evidence.rs:230]: Added comments with links to the pdf

File: verifier/src/lib.rs [hunk 5]
```diff
@@ -49,7 +49,44 @@ pub use x509::{
...
+    ///   from the leaf certificate.
+    /// * [`Error::SignatureVerification`] - If there is a failure verifying one of the signatures
+    ///   in the `certificate_chain`.
+    /// * [`Error::GeneralCertificateError`] - Any other error verifying the `certificate_chain`
+    ///   that isn't covered by the other errors.
+    fn verify_certificate_chain<'a, 'b>(
+        &self,
+        certificate_chain: impl IntoIterator<Item = &'a Certificate>,
+        crls: impl IntoIterator<Item = &'b CertificateList>,
+        _time: DateTime,
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply verifier/src/lib.rs]: fixed thanks!

File: verifier/src/lib.rs [hunk 6]
```diff
@@ -49,7 +49,44 @@ pub use x509::{
...
+/// trust anchor.
+pub trait CertificateChainVerifier {
+    /// Verify a certificate chain.
+    ///
+    /// # Returns
+    /// The public signing key from the leaf certificate.
+    ///
+    /// # Arguments
+    /// * `certificate_chain` - The certificate chain to verify.
+    /// * `crls` - The certificate revocation lists to use when verifying the certificate chain.
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply verifier/src/lib.rs]: fixed thanks!

File: verifier/src/error.rs [hunk 7]
```diff
@@ -39,6 +39,16 @@ pub enum Error {
...
+    /// X509 certificate not yet valid
+    CertificateNotYetValid,
+    /// X509 certificate has expired
+    CertificateExpired,
+    /// X509 certificate has been revoked
+    CertificateRevoked,
+    /// Error decoding a public key
+    PublicKeyDecodeError,
+    /// General error trying to verify a certificate chain
+    GeneralCertificateError,
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply verifier/src/error.rs]: Good suggestion! The error usage is more or less isolated. 
Updated to dedicated error type,

**Comment by nick-mobilecoin** [comment]
@eranrund I marked for re-review since it was 100+ lines changed in addressing the suggestions.

**Review by eranrund — APPROVED** [review]

**Pull request merged**